File: api/src/main/java/org/zenika/api/service/PokemonService.java
package org.zenika.api.service;

import org.springframework.stereotype.Service;
import org.zenika.api.client.ArenaApi;
import org.zenika.core.Battle;
import org.zenika.core.Pokemon;
import org.zenika.api.repository.PokemonRepository;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.Random;
import java.util.stream.Collectors;

@Service
public class PokemonService {

    private PokemonRepository pokemonRepository;
    private ArenaApi arenaApi;

    public PokemonService(PokemonRepository pokemonRepository, ArenaApi arenaApi) {
        this.pokemonRepository = pokemonRepository;
        this.arenaApi = arenaApi;
    }

    public List<Pokemon> getAllPokemon() {
        return pokemonRepository.findAll().stream().map(pokemon -> pokemon.toPokemon()).collect(Collectors.toList());
    }

    public Pokemon getPokemonByIdOrName(String idOrName) {
        final Optional<org.zenika.api.entity.Pokemon> pokemonByName = pokemonRepository.findByName(idOrName);
        if (pokemonByName.isPresent()) {
            return pokemonByName.get().toPokemon();
        }

        Optional<org.zenika.api.entity.Pokemon> pokemonById = pokemonRepository.findById(Long.parseLong(idOrName));
        if (pokemonById.isPresent()) {
            return pokemonById.get().toPokemon();
        }
        return null;
    }

    public List<String> getAllPokemonTypes() {
        return getAllPokemon().stream()
                .map(it -> it.getTypes())
                .flatMap(List::stream)
                .distinct()
                .collect(Collectors.toList());
    }

    public Battle fight(String idPokemon1, String idPokemon2) {
        Pokemon pokemon1 = getPokemonByIdOrName(idPokemon1);
        Pokemon pokemon2 = getPokemonByIdOrName(idPokemon2);

        if (pokemon1 == null || pokemon2 == null) {
            return null;
        }
        return arenaApi.fight(Arrays.asList(pokemon1, pokemon2));
    }

    public Battle getBattle(String id) {
        return arenaApi.get(id);
    }
}

File: api/src/main/kotlin/org/zenika/api/service/PokemonService.kt
package org.zenika.api.service;

import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service;
import org.zenika.api.client.ArenaApi;
import org.zenika.api.entity.BattleEntity
import org.zenika.core.Battle;
import org.zenika.core.Pokemon;
import org.zenika.api.repository.PokemonRepository;

@Service
class PokemonService(val pokemonRepository: PokemonRepository, val arenaApi: ArenaApi) {

    fun getAllPokemon(): List<Pokemon> = pokemonRepository.findAll().map { it.toPokemon() }

    fun getPokemonByIdOrName(idOrName: String): Pokemon? {
        val pokemonByName = pokemonRepository.findByName(idOrName)
        if (pokemonByName != null) {
            return pokemonByName.toPokemon()
        }
        return pokemonRepository.findByIdOrNull(idOrName.toLong())?.toPokemon()
    }

    fun getAllPokemonTypes(): List<String> = getAllPokemon().flatMap { it.types }.distinct()

    fun fight(idPokemon1: String, idPokemon2: String): Battle? {
        val pokemon1 = getPokemonByIdOrName(idPokemon1)
        val pokemon2 = getPokemonByIdOrName(idPokemon2)
        if (pokemon1 == null && pokemon2 == null) {
            return null
        }
        return arenaApi.fight(listOfNotNull(pokemon1, pokemon2))
    }

    fun getBattle(id: String) = arenaApi.get(id)
}
